x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/PushTopNToSource.java

```diff
@@ -7,55 +7,212 @@
     protected PhysicalPlan rule(TopNExec topNExec, LocalPhysicalOptimizerContext ctx) {
-        PhysicalPlan plan = topNExec;
-        PhysicalPlan child = topNExec.child();
-        if (canPushSorts(child)
-            && canPushDownOrders(topNExec.order(), x -> LucenePushDownUtils.hasIdenticalDelegate(x, ctx.searchStats()))) {
+        Pushable pushable = evaluatePushable(topNExec, x -> LucenePushDownUtils.hasIdenticalDelegate(x, ctx.searchStats()));
+        return pushable.rewrite(topNExec);
+    }
+
+    /**
+     * Multiple scenarios for pushing down TopN to Lucene source. Each involve checking a combination of conditions and then
+     * performing an associated rewrite specific to that scenario. This interface should be extended by each scenario, and
+     * include the appropriate rewrite logic.
+     */
+    interface Pushable {
+        PhysicalPlan rewrite(TopNExec topNExec);
+    }
+
+    private static final Pushable NO_OP = new NoOpPushable();
+
+    record NoOpPushable() implements Pushable {
+        public PhysicalPlan rewrite(TopNExec topNExec) {
+            return topNExec;
+        }
+    }
+
+    record PushableExchangeExec(ExchangeExec exchangeExec, EsQueryExec queryExec) implements Pushable {
+        public PhysicalPlan rewrite(TopNExec topNExec) {
+            var sorts = buildFieldSorts(topNExec.order());
+            var limit = topNExec.limit();
+            return exchangeExec.replaceChild(queryExec.withSorts(sorts).withLimit(limit));
+        }
+    }
+
+    record PushableQueryExec(EsQueryExec queryExec) implements Pushable {
+        public PhysicalPlan rewrite(TopNExec topNExec) {
             var sorts = buildFieldSorts(topNExec.order());
             var limit = topNExec.limit();
+            return queryExec.withSorts(sorts).withLimit(limit);
+        }
+    }
+
+    record PushableGeoDistance(FieldAttribute fieldAttribute, Order order, Point point) {
+        private EsQueryExec.Sort sort() {
+            return new EsQueryExec.GeoDistanceSort(fieldAttribute.exactAttribute(), order.direction(), point.getLat(), point.getLon());
+        }
 
-            if (child instanceof ExchangeExec exchangeExec && exchangeExec.child() instanceof EsQueryExec queryExec) {
-                plan = exchangeExec.replaceChild(queryExec.withSorts(sorts).withLimit(limit));
-            } else {
-                plan = ((EsQueryExec) child).withSorts(sorts).withLimit(limit);
+        private static PushableGeoDistance from(StDistance distance, Order order) {
+            if (distance.left() instanceof Attribute attr && distance.right().foldable()) {
+                return from(attr, distance.right(), order);
+            } else if (distance.right() instanceof Attribute attr && distance.left().foldable()) {
+                return from(attr, distance.left(), order);
             }
+            return null;
+        }
+
+        private static PushableGeoDistance from(Attribute attr, Expression literal, Order order) {
```

Comment:
nit:
```suggestion
        private static PushableGeoDistance from(Attribute attr, Expression foldable, Order order) {
```
otherwise this suggests that only `Literal` instances should be used, but in fact foldability is all we need.

Similarly, I think `makeGeometryFromLiteral` should be renamed to `makeGeometryFromFoldable`.